src/dataflow/source/kafka.rs

```diff
@@ -415,44 +426,45 @@ fn downgrade_capability(
-                            consumer,
-                            topic,
-                            *partition_count,
-                            last_processed_offset,
-                            next_partition_ts,
-                            *last_closed_ts,
-                            start_offset,
-                        );
-                        *current_partition_count = i32::try_from(partitions.len()).unwrap();
-                    }
-                    if last_offset == *offset {
-                        // We have now seen all messages corresponding to this timestamp for this
-                        // partition. We
-                        // can close the timestamp (on this partition) and remove the associated metadata
-                        next_partition_ts.insert(pid.clone(), *ts);
-                        entries.remove(0);
-                        changed = true;
-                    } else {
-                        // Offset isn't at a timestamp boundary, we take no action
-                        break;
-                    }
+    if let Some(entries) = timestamp_histories.borrow_mut().get_mut(id) {
+        for (pid, entries) in entries {
+            // Obtain the last offset processed (or 0 if no messages have yet been processed)
+            let last_offset = *last_processed_offset.get(pid).unwrap_or(&0);
+            // Check whether timestamps can be closed on this partition
+            while let Some((partition_count, ts, offset)) = entries.first() {
+                assert!(*offset >= start_offset, "Internal error! Timestamping offset went below start: {} < {}. Materialize will now crash.", offset, start_offset);
+                if partition_count > current_partition_count {
+                    // A new partition has been added, we need to update the appropriate
+                    // entries before we continue. This will also update the last_processed_offset
+                    // and next_partition_ts data structures
+                    let partitions = update_partition_list(
+                        consumer,
+                        topic,
+                        *partition_count,
+                        last_processed_offset,
+                        next_partition_ts,
+                        *last_closed_ts,
+                        start_offset,
+                    );
+                    *current_partition_count = i32::try_from(partitions.len()).unwrap();
+                }
+                if last_offset == *offset {
+                    // We have now seen all messages corresponding to this timestamp for this
+                    // partition. We
+                    // can close the timestamp (on this partition) and remove the associated metadata
+                    next_partition_ts.insert(pid.clone(), *ts);
+                    entries.remove(0);
+                    changed = true;
+                    KAFKA_PARTITION_OFFSET_INGESTED
```

Comment:
I don't think KAFKA_PARTITION_OFFSET_INGESTED should be here. This is when we close the timestamp. 

There's three steps
1) ingest the message (this happens when we call consumer.poll()
2) timestamp the message and "give it" to Timely. This happens when we call                             output.session(&ts_cap).give((key, (out, Some(offset - 1))));
3) close out the timestamp and upgrade the capability (this happens when you call cap.downgrade()).

The place where you put it is where we close out a timestamp on a particular partition. I think putting it either at 1) or at 2) makes the most sense. The difference between 1 and 2 shows how much our timestamping is lagging behind our ingestion.

Reply:
interesting I was thinking of `cap.downgrade` as being when the data is "fully ingested" -- like this is the soonest moment that the downstream operator can start processing it.

Reply:
I'd be concerned about updating the metric on every message. Does it cause non-trivial work to be done? Or is prometheus good about buffering things

Reply:
This is just a hashmap lookup + atomic integer set, no network traffic happens. If we extracted the `with_label_values` call we could get rid of the hash lookup.

Reply:
It's the soonest moment at which we start processing the message, but it's technically not when we've ingested it.

Like for instance, if we decided to timestamp records in batch 1 million, we'd show the graph of us trailing high_watermarks by a million because we'd only ever be downgrading the capability every million operation. I think that's fine, but it seems to be a slightly different metric than when we actually finished reading the data from Kafka.

Reply:
@quodlibetor hashmap lookup + atomic integer set is insignificant compared to the amount of work we have to do per message anyway, so I retract that particular concern.

Reply:
ahhh that's legit @nacrooks , I agree with your interpretation. Thanks!